pennylane/transforms/decompositions/single_qubit_unitary.py

```diff
@@ -261,6 +261,76 @@ def _xyx_decomposition(U, wire, return_global_phase=False):
     return operations
 
 
+def _xzx_decomposition(U, wire, return_global_phase=False):
+    r"""Computes the decomposition of a single-qubit matrix :math:`U` in terms
+    of elementary operations, as a product of Z and X rotations in the form
+    :math:`e^{i\gamma} RX(\phi) RZ(\theta) RX(\lambda)`. (batched operation)
+
+    Args:
+        U (tensor): A :math:`2 \times 2` unitary matrix.
+        wire (Union[Wires, Sequence[int] or int]): The wire on which to apply the operation.
+        return_global_phase (bool): Whether to return the global phase ``qml.GlobalPhase(-alpha)``
+            as the last element of the returned list of operations.
+
+    Returns:
+        list[Operation]: Returns a list of gates, an ``RX``, an ``RZ`` and
+        another ``RX`` gate, which when applied in the order of appearance in the list is
+        equivalent to the unitary :math:`U` up to a global phase. If `return_global_phase=True`,
+        the global phase is returned as the last element of the list.
+
+    **Example**
+
+    >>> U = np.array(
+    ...     [[-0.28829348-0.78829734j,  0.30364367+0.45085995j],
+    ...      [ 0.53396245-0.10177564j,  0.76279558-0.35024096j]]
+    ... )
+    >>> decompositions = _xzx_decomposition(U, 0, return_global_phase=True)
+    >>> decompositions
+    [RX(12.416147693665032, wires=[0]),
+     RZ(1.3974974090935608, wires=[0]),
+     RX(11.448040119199066, wires=[0]),
+     GlobalPhase((1.1759220332464762-0j), wires=[0])]
+
+    """
+
+    # Small number to add to denominators to avoid division by zero
+    EPS = 1e-64
+
+    # Choose gamma such that exp(-i*gamma)*U is special unitary (detU==1).
+    U = math.expand_dims(U, axis=0) if len(U.shape) == 2 else U
+    U_det1, gammas = _convert_to_su2(U, return_global_phase=return_global_phase)
+
+    # Compute \phi, \theta and \lambda after analytically solving for them from
+    # U_det1 = RX(\phi) RZ(\theta) RX(\lambda)
+    sum_diagonal = math.real(U_det1[:, 0, 0] + U_det1[:, 1, 1])
+    sum_off_diagonal_imag = math.imag(U_det1[:, 0, 1] + U_det1[:, 1, 0])
+    phi_plus_lambdas_d2 = math.arctan2(-sum_off_diagonal_imag, sum_diagonal + EPS)
+    diff_diagonal_imag = math.imag(U_det1[:, 0, 0] - U_det1[:, 1, 1])
+    diff_off_diagonal = math.real(U_det1[:, 0, 1] - U_det1[:, 1, 0])
+    phi_minus_lambdas_d2 = math.arctan2(diff_off_diagonal, -diff_diagonal_imag + EPS)
+    lams = phi_plus_lambdas_d2 - phi_minus_lambdas_d2
+    phis = phi_plus_lambdas_d2 + phi_minus_lambdas_d2
+
+    # Compute \theta
+    thetas = math.where(
```

Comment:
Could you remind me why is the `where` necessary?

Reply:
Basically the diagonal elements add up to
$$\cos\left(\frac{\phi}{2}+\frac{\lambda}{2}\right)\cdot 2\cos\left(\frac{\theta}{2}\right)$$
and the off diagonal elements add up to
$$-\sin\left(\frac{\phi}{2}+\frac{\lambda}{2}\right)\cdot 2i\sin\left(\frac{\theta}{2}\right)$$
Either of them can be used to calculate $\theta$ given $\phi$ and $\lambda$, but if either $\cos(\phi/2+\lambda/2)=0$ or $\sin(\phi/2+\lambda/2)=0$, we must use the other one, or else we're dealing with $0/0$.

If you're asking why we use ```where``` instead of a simple ```if``` statement, this is how ```_xyx_decomposition``` does it, the reason is that there might be multiple matrices come in at the same time, for example, when the unitary is ```qml.Rot(np.array([1.2, 1.5]), np.array([1.2, 1.5]), np.array([1.2, 1.5]), wires=0).matrix()``` like in the test case.

Reply:
Thank you for the detailed answer, that's great.